documentation/typescript/adding-custom-login-form-with-spring-security.asciidoc

```diff
@@ -0,0 +1,176 @@
+
+  constructor(){
+    super();
+    this.onSuccess = () => {
+      Router.go(this.returnUrl);
+    };
+  }
+
+  render() {
+    return html`
+      <vaadin-login-overlay
+        ?opened="${this.open}" 
+        .error=${this.error}
+        @login="${this.login}">    
+      </vaadin-login-overlay>
+    `;
+  }
+
+  async login(event: CustomEvent): Promise<LoginResult> {
+    this.error = false;
+    // use the Vaadin provided login helper method to obtain the LoginResult
+    const result = await login(event.detail.username, event.detail.password);
+    this.error = result.error;
+  
+    if (!result.error) {
+      this.onSuccess(result);
+    }
+
+    return result;
+  }
+
+  onAfterEnter(location: RouterLocation) {
+    this.returnUrl = location.redirectFrom || this.returnUrl;
+  }
+}
+----
+After the login view is defined, you should also define a route for the login view component, for example in the `index.ts` file. Note, the `path` for the login view component should be the same as what you defined in `SecurityConfig` for `http.formLogin().loginPage()`.
+.frontend/index.ts
+[source, typescript]
+----
+const routes = [
+  {
+    path: '/login',
+    component: 'login-view'
+  },
+  // more routes
+}
+----
+== Redirect the user to the login view
+To protect a view from unauthenticated users, that is, redirect unauthenticated users to the login view, you can use route action.
+
+.frontend/index.ts
+[source, typescript]
+----
+const isUserLoggedIn = function() {
+  return localStorage.getItem('loggedIn') === 'true';
+}
+
+const routes = [
+  ...
+  {
+    path: '/my-view',
+    action: async (_: Router.Context, commands: Router.Commands) => {
+      if (!await isUserLoggedIn()) {
+        return commands.redirect('/login');
+      }
+      return undefined;
+    },
+    component: 'my-view'
+  }
+  ...
+}
+----
+You can also add the route action to the parent layout, so that all the child views are protected.
+
+.frontend/index.ts
+[source, typescript]
+----
+const routes = [
+  ...
+  {
+    path: '/main-layout',
+    action: async (_: Router.Context, commands: Router.Commands) => {
+      if (!await isUserLoggedIn()) {
+        return commands.redirect('/login');
+      }
+      return undefined;
+    },
+    component: 'main-layout',
+    children: [
+      ...
+    ]
+  }
+  ...
+}
+----
+The `isUserLoggedIn()` method in the above code examples uses
+a `loggedIn` boolean stored in the https://vaadin.com/components/browse[localstorage] to check if the user is logged in. The boolean needs to be reset when logging out.
+
+
+== Logout
+There are two ways to implement logout. You could either:
+
+Let Spring Security handle Logging out. How it works is that it handles HTTP GET requests to `/logout` and redirects to `/login?logout` in response.
+For that a "Logout" button should be an regular <a> tag, that is,
+`<a href="/logout" router-ignore>Log out</a>`. The downside is that it requires a full page load.
```

Comment:
I don't think this works with the localStorage implementation anymore, since this way it would not clear the "loggedIn" status and you would still be able to see the views even after navigating to "/logout". I think this part can probably be just removed. Or if we want to keep this as an option, we would probably need to add some code that would update the "loggedIn" status in localStorage by checking it from the server on page load (reload) **if** the browser is online.

Reply:
If this is removed, also update the "There are two ways to implement logout. You could either:" above since there would only be one example anymore.

Reply:
good catch, removed the content.